staking.md

```diff
@@ -0,0 +1,285 @@
+- **Locked Balance (LB)**: The total amount of tokens locked in the current staking cycle, i.e., epochs `N` to `N + 3`.
+- **Available Balance (AB)**: Calculated as the difference between TB and LB.
+  - **AB = TB - LB**
+
+### **Lifecycle of Tokens in the Staking Contract**
+
+1. **Staking Tokens (`deposit()`):**
+   - Increases TB and AB.
+   - No immediate effect on staking commitments until `adjustStake()` (or `stake()`) is called.
+
+2. **Adjusting Stake (`adjustStake()`):**
+   - Adjusts the participant's stake commitment for the next staking cycle.
+   - **Positive Adjustment:**
+     - Decreases AB by `amount`.
+     - Increases the stake commitment for the next staking cycle by `amount`.
+   - **Negative Adjustment:**
+     - Increases AB by `-amount`.
+     - Decreases the stake commitment for the next staking cycle by `-amount`.
+   - **AB Recalculation:**
+     - **AB = TB - LB**
+
+3. **Withdrawal (`withdraw()`):**
+   - Decreases TB and AB by `amount`.
+   - Cannot affect tokens locked in the any of the 4 concurrent staking cycles.
+
+4. **Epoch Transition:**
+   - At the end of the Payout Epoch (`N + 3`):
+     - Tokens locked in the current staking cycle become available.
+     - **LB** decreases by the amount of tokens that were locked.
+     - **AB** increases accordingly.
+
+---
+
+### **Updated Example Scenario**
+
+Let's illustrate how the balances change over epochs with this updated model.
+
+#### **Constants and Parameters**
+
+- **Current Epoch (`N`)**: 10 (Staking Epoch)
+- **Staking Cycle Length (`C`)**: 4 epochs (epochs 10 to 13)
+  - Epoch 10: Staking Epoch
+  - Epoch 11: Leader Election Epoch
+  - Epoch 12: Block Production Epoch
+  - Epoch 13: Payout Epoch
+- **Participant**: Alice
+- **`MINIMUM_STAKE`**: 100 tokens
+
+#### **Alice's Balances**
+
+- **Total Balance (TB)**: 1000 tokens (Alice has already deposited these via `deposit()`).
+- **Available Balance (AB)**: 1000 tokens (initially, no tokens are locked).
+- **Locked Balance (LB)**: 0 tokens.
+
+#### **Epoch 10 (Staking Epoch)**
+
+- **Action 1**: Alice decides to stake 400 tokens for the next staking cycle (epochs 11 to 13).
+  - She calls `adjustStake(400)`.
+  - **AB** decreases by 400 tokens:
+    - AB = 1000 - 400 = 600 tokens.
+  - **Stake Commitment**:
+    - Alice's stake commitment for the next staking cycle is now 400 tokens.
+
+- **Action 2**: Alice withdraws 100 tokens.
+  - She calls `withdraw(100)`.
+  - **TB** decreases by 100 tokens:
+    - TB = 1000 - 100 = 900 tokens.
+  - **AB** decreases by 100 tokens:
+    - AB = 600 - 100 = 500 tokens.
+
+- **Locked Balance (LB)**:
+  - Set at 400 from the call to adjustStake till the end of the Staking Epoch.
+
+#### **Epoch 11 (Leader Election Epoch)**
+
+- **Transition to New Staking Cycle**:
+  - Alice's stake of 400 tokens becomes locked for the duration of the staking cycle (epochs 11 to 13).
+  - **LB** increases by 400 tokens:
```

Comment:
Coming from the question to the staking schedule: Is something imperatively setting this value, or is it  calculated by whatever entity that needs it? This would mean we would have to iterate over all the stakings this user ever did to find the latest, relevant ones - or do we have efficient ways to access only the 'last couple' elements of a list?